Update MSFT_IntuneAppProtectionPolicyiOS.psm1 - fixes #2942

menswearUK opened the pull request:
added New-M365DSCConnection command to address #2942 - added formatting to current/target values output

#### Pull Request (PR) description
<!--
    Replace this comment block with a description of your PR.
-->

This PR adds a line:
$ConnectionMode = New-M365DSCConnection -Workload 'MicrosoftGraph' -ProfileName beta `
    -InboundParameters $PSBoundParameters

to the Test-TargetResource function of IntuneAppProtectionPolicyiOS

It prevents the Get-TargetResource function returning objects with missing values.  It's not clear to me why the same line in get-TargetResource doesn't work correctly but without this line added here the object returned from the MS graph command uses the v1.0 profile and is missing a number of values which are required for the returned array.  This leads to failed comparisons with config documents even when the settings are correct.

The error and the fix can be seen in a verbose output of start-DSCConfiguration or in test-DSCConfiguration

#### This Pull Request (PR) fixes the following issues
<!--
    If this PR does not fix an open issue, replace this comment block with None.
    If this PR resolves one or more open issues, replace this comment block with
    a list the issues using a GitHub closing keyword, e.g.:
    - Fixes #123
    - Fixes #124
-->
Fixes #2942

ykuijs commented:
Even though this solution works, it is more like a workaround. I rather try to solve the underlying issue before implementing a workaround.

menswearUK commented:
I thought it was easiest to update this PR with the places I added code into New-M365DscConnection - since the PR is on hold and, once we reach a solution, I can amend further as necessary.
In addition to my initial suggestion I also added a check that the New-M365DscConnection was for MSGraph before applying the switch as it could be used for multiple connections.
In my testing there was no requirement to add further verbose settings changes if the command is run within New-M365DscConnection but I can add those in too if they are required

menswearUK commented:
@ykuijs  - I've updated this PR with the fixes discussed in the issue chat - the only thing I've not done is added in the changes to the verbose output (it only seems to be an issue when the profile switch is in the configuration, it doesn't seem to be required in this function although I can add those lines in if you'd prefer).

ykuijs commented on CHANGELOG.md:
```diff
@@ -1,6 +1,9 @@
 # Change log for Microsoft365DSC
 
 # UNRELEASED
+* New-M365DSCConnection in M365DSCUtil.psm1
```
Can you please move this item under the MISC section (currently on line 14)

ykuijs commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppProtectionPolicyiOS/MSFT_IntuneAppProtectionPolicyiOS.psm1:
```diff
@@ -1080,8 +1080,8 @@ function Test-TargetResource
 
-    Write-Verbose -Message "Current Values: $(Convert-M365DscHashtableToString -Hashtable $CurrentValues)"
-    Write-Verbose -Message "Target Values: $(Convert-M365DscHashtableToString -Hashtable $PSBoundParameters)"
+    Write-Verbose -Message "Current Values: $((Convert-M365DscHashtableToString -Hashtable $CurrentValues) -replace ';', "`r`n")"
```
Can you leave this as it was? I like the change, but now it is only implemented for this resource and not for all other resources. I rather change the Convert-M365DscHashtableToString function to do this, so all resources replace the semi colon with an enter.

Please create an issue for this so we can assess the impact and implement separately.

ykuijs commented on Modules/Microsoft365DSC/Modules/M365DSCUtil.psm1:
```diff
@@ -1575,6 +1575,13 @@ function New-M365DSCConnection
+            #2942 - check correct profile applied
+            if (($ProfileName -ne $global:MSCloudLoginConnectionProfile.MicrosoftGraph.ProfileName) -and ($Workload -eq 'MicrosoftGraph') )
+            {
+                select-mgprofile -Name $profilename
```
Could you update the cmdlet to match the default PascalCase format: Select-MgProfile

And update the variable name to match the earlier used PascalCasing: $ProfileName

Both are true for all instances you have added.

ykuijs reviewed (changes requested):
Thanks for this update. Just a few minor comments.

menswearUK commented:
@ykuijs No worries, I've corrected those values now.

ykuijs commented:
@menswearUK, thanks for that. But what about the other two comments?

ykuijs reviewed (approved):
LGTM

ykuijs commented:
@menswearUK, wanted to merge it, so I could include it in this weeks release. But unfortunately I was unable to. Could you update the PR with the Dev branch?

menswearUK commented:
@ykuijs - apologies for this, I was attempting to clear the conflicts and messed it up - I ended up clearing the commits and syncing and it closed the PR with no changes.  I've created a new branch with the updates which has no conflicts and I've put in  a new PR for that.
